kube-client/src/api/portforward.rs

```diff
@@ -255,22 +264,27 @@ async fn forwarder_loop<S>(
 where
     S: AsyncRead + AsyncWrite + Unpin + Sized + Send + 'static,
 {
-    // Keep track if the channel has received the initialization frame.
-    let mut initialized = vec![false; 2 * ports.len()];
-    let mut shutdown = vec![false; 2 * ports.len()];
+    #[derive(Default, Clone)]
+    struct ChannelState {
+        // Keep track if the channel has received the initialization frame.
+        initialized: bool,
+        // Keep track if the channel has shutdown.
+        shutdown: bool,
+    }
+    let mut chan_state = vec![ChannelState::default(); 2 * ports.len()];
     let mut closed_ports = 0;
     let mut socket_shutdown = false;
     while let Some(msg) = receiver.next().await {
         match msg {
             Message::FromPod(ch, mut bytes) => {
                 let ch = ch as usize;
-                if ch >= initialized.len() {
+                if ch >= chan_state.len() {
                     return Err(Error::InvalidChannel(ch));
                 }
```

Comment:
I think this could be:

```rust
let ch = chan_state.get_mut(usize::from(ch)).ok_or(Error::InvalidChannel(ch))?;
```

Which should allow us to avoid constantly reindexing into the array everywhere.

I'm also a bit confused about why we're always using `port_index` for `shutdown` but `ch` for `initialized`.

Reply:
Is this because we need to track initialization per channel but shutdown per (two-channel) port?

Reply:
Yeah so we only send shutdown to the writers half, which is the same index as the the even `ch`.

Reply:
Wouldn't that mean that we're mixing up the states as currently implemented though?

We'd be writing "port 1 (chan 2) has been shut down" into the state for chan 1 (port 0's error channel).

Reply:
That would probably still _work_ (for our current purposes anyway), but it'd be quite confusing to try to debug later on.

Reply:
> We'd be writing "port 1 (chan 2) has been shut down" into the state for chan 1 (port 0's error channel).

You mean when we receive `ToPodClose` ?
hmm, we only call shutdown via index when receiving `ToPodClose` and `ToPodClose` is only sent to the "data" channels, because `to_pod_loop` only runs on the "data" channels?

I wonder if we can simplify this by having a `FromPodError` so we don't need a fake channel for the `error_senders`, what do you think?

Reply:
> You mean when we receive ToPodClose ?
Yes.
> I wonder if we can simplify this by having a FromPodError so we don't need a fake channel for the error_senders, what do you think?
We can't really get rid of it entirely from the internals, since it's a Kubernetes-API-ism. I would like to get away from library consumers having to consume a separate stream for the errors, but I believe that would probably require us to switch from `AsyncRead` to `TryStream`, which would be a larger and breaking change for the users.

Shorter term, I _think_ it should be ok to just use `chan_state[ch].shutdown` rather than `chan_state[port_index].shutdown`, which should at least ensure that we store it together with the correct `initialized`. Even if `chan_state[ch].shutdown` is perpetually `false` for error channels.

Reply:
Yep, I'll make that change.
Btw, this is sort of what I was thinking about: https://gist.github.com/tiagolobocastro/9184e000f5bd1116313968bba8e7a935, but not sure if it actually clears things up or end up confusing further :disappointed:

Reply:
I think that might make sense as a future followup PR, or at least be an avenue worth investigating.

Reply:
Now that we use `chan_state[ch]` consistently, would you mind making the change suggested at the start of the thread?

Reply:
Ah yes, done